### PR TITLE
Expose the new gizmo plugin system to scripting

### LINKED ISSUE #23271: Plugins: Ability to create custom EditorSpatialGizmo
**Godot version:**
Godot 3.1  df3fe9d78

**OS/device including version:**
Windows 10

**Issue description:**
In Godot 3.0 was possible to create custom gizmo from GDScript plugin by override [EditorPlugin.create_spatial_gizmo](https://docs.godotengine.org/en/3.0/classes/class_editorplugin.html#class-editorplugin-create-spatial-gizmo)
In the latest Godot this functionality was removed and currently no way to create custom spatial gizmo from GDScript

### PR DESCRIPTION
**Pull request by JFonS**
I will write some documentation for this once it's merged and I have some free time :)

Closes #23271

### CODE REVIEW

File: editor/plugins/spatial_editor_plugin.cpp [hunk 1]
```diff
@@ -5730,8 +5731,16 @@ void SpatialEditorPlugin::snap_cursor_to_plane(const Plane &p_plane) {
 	spatial_editor->snap_cursor_to_plane(p_plane);
 }
 
-void SpatialEditor::register_gizmo_plugin(Ref<EditorSpatialGizmoPlugin> ref) {
+void SpatialEditor::add_gizmo_plugin(Ref<EditorSpatialGizmoPlugin> ref) {
```
Comment (akien-mga) [inline editor/plugins/spatial_editor_plugin.cpp]: Should be `p_ref` according to our convention, though IMO `p_plugin` might be clearer.

File: editor/plugins/spatial_editor_plugin.cpp [hunk 2]
```diff
@@ -5730,8 +5731,16 @@ void SpatialEditorPlugin::snap_cursor_to_plane(const Plane &p_plane) {
...
+	ERR_FAIL_NULL(ref.ptr())
 	gizmo_plugins.push_back(ref);
+	_update_gizmos_menu();
+	SpatialEditor::get_singleton()->update_all_gizmos();
+}
+
+void SpatialEditor::remove_gizmo_plugin(Ref<EditorSpatialGizmoPlugin> ref) {
```
Comment (akien-mga) [inline editor/plugins/spatial_editor_plugin.cpp]: And same here.

**Comment by akien-mga** [comment]
Note: labelled "bug" as not having this feature was a regression compared to 3.0.

File: editor/plugins/spatial_editor_plugin.cpp [hunk 3]
```diff
@@ -5890,22 +5910,109 @@ Ref<EditorSpatialGizmo> EditorSpatialGizmoPlugin::get_gizmo(Spatial *p_spatial)
 	return ref;
 }
 
+void EditorSpatialGizmoPlugin::_bind_methods() {
+#define GIZMO_REF PropertyInfo(Variant::OBJECT, "gizmo", PROPERTY_HINT_RESOURCE_TYPE, "EditorSpatialGizmo")
+
+	BIND_VMETHOD(MethodInfo(Variant::BOOL, "has_gizmo", PropertyInfo(Variant::OBJECT, "spatial", PROPERTY_HINT_RESOURCE_TYPE, "Spatial")));
+	BIND_VMETHOD(MethodInfo(GIZMO_REF, "has_gizmo", PropertyInfo(Variant::OBJECT, "spatial", PROPERTY_HINT_RESOURCE_TYPE, "Spatial")));
```
Comment (akien-mga) [inline editor/plugins/spatial_editor_plugin.cpp]: Does it work to use a detailed `PropertyInfo` for the return type? Or is it the same as `Variant::OBJECT` in the end?

Comment (akien-mga) [reply editor/plugins/spatial_editor_plugin.cpp]: Also is it expected to have two vmethods named `has_gizmo` with different return type?

Comment (akien-mga) [reply editor/plugins/spatial_editor_plugin.cpp]: Looking at the code, I think it was meant to be `create_gizmo` here.

Comment (akien-mga) [reply editor/plugins/spatial_editor_plugin.cpp]: To rule out potential issues, might be worth running `godot --doctool .` to generate the empty template, even if you'll only fill it later.

**Comment by Zylann** [comment]
Would this help to fix #17499?

**Comment by akien-mga** [comment]
X11 and server `tools=yes` fail building:
```
scene/3d/spatial.cpp: In member function 'void Spatial::update_gizmo()':
scene/3d/spatial.cpp:404:43: error: 'SceneTree::GroupCallFlags' is not a class or namespace
   get_tree()->call_group_flags(SceneTree::GroupCallFlags::GROUP_CALL_REALTIME, SceneStringNames::get_singleton()->_spatial_editor_group, SceneStringNames::get_singleton()->_request_gizmo, this);
                                           ^
```

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**